[6.0] Configure Joomla with environment variables

voronkovich opened the pull request:
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/36898.

### Summary of Changes

This PR adds an ability to configure Joomla using environment variables. To do this, the config provider has been changed to load all environment variables whose names start with `JOOMLA_` (e.g. `JOOMLA_DEBUG`, `JOOMLA_PASSWORD`) and merge them into the `config` service.

Also, environment variables can be loaded from `.env` files, thankfully to [symfony/dotenv](https://github.com/symfony/dotenv).

### Testing Instructions

1. Create `.env` file in the site's root directory

2. Set environment variables (inside the `.env` file):
   ```env
   JOOMLA_OFFLINE=1
   JOOMLA_OFFLINE_MESSAGE='Joomla!'
   ```
3. Reload site's home page and ensure that site is offline and the `Joomla!` message is shown

4. Run the following command and ensure that the output is `true`:
   
    ```sh
    cli/joomla.php config:get offline
    ```
5. Run the following command and ensure that the output is `Joomla!`:
   
    ```sh
    cli/joomla.php config:get offline_message
    ```

### Actual result BEFORE applying this Pull Request

All should work.

### Expected result AFTER applying this Pull Request

All should work.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### Why?

The [12-factor app](https://12factor.net) methodology recommends to [store config in the environment](https://12factor.net/config).

Also, it makes Joomla more [cloud-friendly](https://atlan.com/cloud-friendly-vs-cloud-native/#what-makes-an-application-cloud-friendly) and simplifies running it inside containers.


For example, using environment variables it's very easy to configure separate database for testing:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<phpunit bootstrap="tests/Unit/bootstrap.php" colors="false">
        <testsuites>
                <testsuite name="Unit">
                        <directory suffix="Test.php">./tests/Unit</directory>
                </testsuite>
                <testsuite name="Integration">
                        <directory suffix="Test.php">./tests/Integration</directory>
                </testsuite>
        </testsuites>
        <php>
                <env name="JOOMLA_DBHOST" value="localhost" force="true"/>
                <env name="JOOMLA_DBNAME" value="joomla_test" force="true" />
                <env name="JOOMLA_DBUSER" value="joomla_test_user" force="true" />
                <env name="JOOMLA_DBPASSWORD" value="joomla_test_password" force="true" />
        </php>
</phpunit>
```

voronkovich committed:
Configure with environment variables

dgrammatiko commented:
Nice one! One suggestion probably another array with key the JConfig key and value 1 if the value comes from the php file or 0 if the value comes from the env. the reason is that then in the admin configuration the form could check if each input is editable ie coming from the php file (or not ie coming from the env and then add a readonly attribute).

voronkovich commented:
@dgrammatiko, to check if the config value came from environment we can use the `envs_config` service:

```php
/** @var Registry */
$envsConfig = $container->get('envs_config');

if ($envsConfig->has('debug')) {
    // 'debug' came from environment
}
```

laoneo commented:
Thanks for the pr. It's a nice idea. For me it is a bit too complex. Probably moving the [loadEnvs](https://github.com/joomla/joomla-cms/pull/44894/files#diff-d1019ae09bd8399289c707cf429a3bffc94f252010dccdee00468c7670374401R94) function to the Config service provider is enough. No need to introduce a storage class with static functions.

voronkovich commented:
@laoneo, without storage and static methods we can't set `debug`, `error_reporting` and `behind_loadbalanser`, because they are configured during application loading in `includes/framework.php`. So, `JOOMLA_DEBUG`, `JOOMLA_ERROR_REPORTING` wouldn't work.

laoneo commented:
Then load them manually in the respective files.

voronkovich committed:
Simplify

SniperSister committed:
[5.2][system tests] Do not delete tuf metadata (#44866)

joomdonation committed:
Set column alias for BannerTable, fix warning on Save As Copy (#44864)

voronkovich commented:
@laoneo, I've added the changes you requested.

joomla-translation-bot committed:
[5.2] Translation Update (#44919)

alikon committed:
[5.2] PHP Deprecated: trim(): Passing null to parameter #1 ($string) of type string is deprecated (#44876)

laoneo commented on libraries/src/Service/Provider/Config.php:
```diff
@@ -39,19 +39,128 @@ public function register(Container $container)
-                    if (!is_file(JPATH_CONFIGURATION . '/configuration.php')) {
-                        return new Registry();
-                    }
+                    $config = new Registry();
```
Please change this block back to early return.

voronkovich replied:
@laoneo, done.

laoneo replied:
Thanks!

voronkovich committed:
Change block back to early return

voronkovich committed:
Minor CS fix

laoneo commented:
Just an update here, before you spend more time on it. We are discussing this in the maintainers team if it fits into our current setup. There are are already other ways to customize the config, so this would be another one.

tekvishal commented:
I have tested this item :red_circle: unsuccessfully on 2466eae24abcaee27be9d7766344c0db3619a8a0<br /><br />I have tested this item, issue not resolved<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44894">issues.joomla.org/tracker/joomla-cms/44894</a>.</sub>

komalm commented:
I have tested this item :red_circle: unsuccessfully on 2466eae24abcaee27be9d7766344c0db3619a8a0<br /><br />Tested on: Joomla! 5.3-dev
Environment: Joomla! 5.3-dev (fresh setup)
Web Server: Apache
PHP Version: PHP 8.3
Database: MySQL

**Steps Followed for Testing:**
- Applied the patch via Patch Tester.
- Edited .htaccess file and added the environment variable: SetEnv JOOMLA_OFFLINE 1
- Restarted the web server : sudo systemctl restart apache2
- Opened the Joomla! frontend to check if the site was set to offline mode


**Issue Observed:**
- The site did not switch to offline mode.
- No errors were logged in Joomla! or the server error logs.
- Environment variables were not detected by Joomla!

**Expected Behavior:**
- Joomla! should recognize the JOOMLA_OFFLINE environment variable.
- The site should display the "Site is offline" message if the variable is set to 1.

**Actual Behavior:**
- Joomla! continues running normally, ignoring the JOOMLA_OFFLINE setting.
- No difference observed before and after applying the patch.

**Additional Notes:**
- Clearing Joomla! cache and restarting the server did not resolve the issue.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44894">issues.joomla.org/tracker/joomla-cms/44894</a>.</sub>

komalm commented:
I have tested this item :white_check_mark: successfully on 2466eae24abcaee27be9d7766344c0db3619a8a0<br /><br />Sincerely apologize for my previous negative feedback regarding the issue. It was provided before properly applying the patch, and I realize now that it may not have reflected the actual results accurately.

After correctly applying the patch and retesting, here are my updated findings:

Tested on: Joomla! 5.3-dev
Environment: Joomla! 5.3-dev (fresh setup)
Web Server: Apache
PHP Version: PHP 8.3
Database: MySQL

**Steps Followed for Testing:**
- Applied the patch via Patch Tester.
- Edited .htaccess file and added the environment variable: SetEnv JOOMLA_OFFLINE 1
- Restarted the web server : sudo systemctl restart apache2
- Opened the Joomla! frontend to check if the site was set to offline mode

**Issue Observed:**
- The site switch to offline mode.
- No errors were logged in Joomla! or the server error logs.

**Expected Behavior & Actual Behavior are same:**
- Joomla! should recognize the JOOMLA_OFFLINE environment variable.
- The site should display the "Site is offline" message if the variable is set to 1.

**Additional Notes:**
- Clearing Joomla! cache and restarting the server did not resolve the issue.

**Final Result:**
- Patch works as intended; Joomla correctly recognizes the JOOMLA_OFFLINE variable, and the site goes offline as expected. No issues found.



<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44894">issues.joomla.org/tracker/joomla-cms/44894</a>.</sub>

drmenzelit committed:
Add border last element media breadcrumbs (#44995)

heelc29 committed:
[5.2] Disable compat plugin for system tests (#44997)

brianteeman committed:
[cypress] Update Menu.cy.js (#45005)

Commit:
Update contact.xml (#44988)

iteidrm committed:
Redesign carousel implementation to reflect documentation - bug fix (#44488)

heelc29 committed:
[5.3] better use str_starts_with (#44998)

joomdonation committed:
[5.3] Combine multiple unset call into one call (#44983)

* Combine multiple unset call into one call

* Remove unnecessary blank lines

* Move comments to right position

* Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

* Update libraries/src/User/User.php

Co-authored-by: Harald Leithner <leithner@itronic.at>

joomdonation committed:
[5.3] Change isset($var) && multiple times to one isset call (#44982)

* Change isset($var) && multiple times to one call

* Remove redundant checks

* Update components/com_users/src/View/Profile/HtmlView.php

---------

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>
Co-authored-by: Harald Leithner <leithner@itronic.at>

joomdonation committed:
Simplify code using no_unneeded_control_parentheses rule (#44981)

Co-authored-by: Harald Leithner <leithner@itronic.at>

joomdonation committed:
Improve api app code using rector rules (#44975)

Co-authored-by: Harald Leithner <leithner@itronic.at>

joomdonation committed:
[5.3] Change list() to array destruct for libraries code (#44973)

* Change list() to array destruct for libraries code

joomdonation committed:
[5.3] Change list() to array destruct for modules and plugins code (#44970)

* Change list() to array destruct for modules and plugins code

---------

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

joomdonation committed:
Changes unneeded null check to ?? operator for modules and plugins (#44969)

joomdonation committed:
Remove unused variable in try catch block for modules and plugins (#44967)

rdeutz commented:
We discussed this at todays maintainers meeting and we think it is a good idea. And now comes the but. But we want to use an industry standard implementation like “[phpdotenv](https://github.com/vlucas/phpdotenv)” or "https://github.com/symfony/dotenv".

QuyTon committed:
Add frontediting setting (#44978)

joomdonation committed:
Fix MailsHelper::loadTranslationFiles (#44952)

SniperSister committed:
Fix dispatcher ACl checks: ignore any non-display tasks (#44951)

bembelimen committed:
[5.2] Add legacy message to content modules (#44366)

* Add deprecation message to content modules

---------

Co-authored-by: Quy <quy@nomonkeybiz.com>

rdeutz committed:
[5.3] Handle fields event in EmailCloak plugin (#44923)

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Hackwar committed:
[5.3] Category View: Fix meta tag for author (#44903)

richard67 committed:
Composer update php-debugbar to 1.23.6 (#44901)

SniperSister committed:
[5.3] Make mailtemplate information available in mail template layout (#44896)

janschoenherr committed:
[5.3] Replace calls to dirname(__FILE) with __DIR__ (#44882)

janschoenherr committed:
[5.3] Remove unnecessary array internal pointer resets (#44881)

voronkovich committed:
Use symfony/dotenv

joomdonation committed:
Use null coalescing assignment operator for modules and plugins (#44972)

joomdonation committed:
Fix Field Group permission check (#44956)

chmst committed:
[5.3] Fields batch copy category (#44943)

muhme committed:
[5.x][docu] Minor sudo fix (#45031)

The `CYPRESS_CACHE_FOLDER` environment variable must be set within the `sudo` command.

muhme committed:
[Cypress] PHP Backend Notice com_media/Files (#44976)

* [Cypress] PHP Backend Notice com_media/Files.cy.js

Fixing all Joomla backend PHP notices from Cypress test file `api/com_media/Files.cy.js` and doing some refactoring.

Fixed PHP notices are:
```
[Sat Feb 22 18:28:44.019593 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  exif_imagetype(): Error reading from /var/www/html/files/test-image-1.jpg! in /var/www/html/libraries/src/Helper/MediaHelper.php on line 93
[Sat Feb 22 18:28:44.022074 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  getimagesize(): Error reading from /var/www/html/files/test-image-1.jpg! in /var/www/html/libraries/src/Image/Image.php on line 177
[Sat Feb 22 18:28:44.258619 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  exif_imagetype(): Error reading from /var/www/html/files/test-dir/test-image-1-subfolder.jpg! in /var/www/html/libraries/src/Helper/MediaHelper.php on line 93
[Sat Feb 22 18:28:44.259092 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  getimagesize(): Error reading from /var/www/html/files/test-dir/test-image-1-subfolder.jpg! in /var/www/html/libraries/src/Image/Image.php on line 177
[Sat Feb 22 18:28:47.177357 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  exif_imagetype(): Error reading from /var/www/html/files/test-dir/todelete.jpg! in /var/www/html/libraries/src/Helper/MediaHelper.php on line 93
[Sat Feb 22 18:28:47.178099 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  getimagesize(): Error reading from /var/www/html/files/test-dir/todelete.jpg! in /var/www/html/libraries/src/Image/Image.php on line 177
[Sat Feb 22 18:28:47.564807 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  exif_imagetype(): Error reading from /var/www/html/images/test-dir/todelete.jpg! in /var/www/html/libraries/src/Helper/MediaHelper.php on line 93
[Sat Feb 22 18:28:47.565288 2025] [php:notice] [pid 31:tid 31] [client 10.0.0.1:64714] PHP Notice:  getimagesize(): Error reading from /var/www/html/images/test-dir/todelete.jpg! in /var/www/html/libraries/src/Image/Image.php on line 177
```

All 8 x PHP notices are fixed. Instead of using 1-Byte files real images are used from Cypress fixtures.
Please consider there are still 4 x PHP warning to be fixed (I assume they need to be fixed in PHP API backend code) from `api/com_media/Files.cy.js` and more warnings and notices in the overall Joomla System Tests.

Some refactoring:
* Deleted 3 images in `tests/System/data/com_media`, 2 are not used and Cypress default is to use `fixtures` folder
* The image file names are counted through and the images show name and path in different colors.
* Second `test-dir` was named `test-dir2` to distinguish clearly.
* `afterEach()` is reduced to `after()`, as it is only a clean-up and not needed to execute the tests.
* The last 4 tests can use the prepared files too, as with `beforeEach` the files are restored before each test.

* Corrected Image Description to test-dir2

Used command:
```
convert -size 100x100 -gravity center -pointsize 11 xc:lightgreen -fill red -annotate +0+0 'Joomla\nSystem Tests\n\nimages/test-dir2\ntest-image-3.jpg' tests/System/fixtures/com_media/test-image-3.jpg
```

* Fix for Windows Path and Read-Only Overwrites

- Creating relative fixtures folder path now also works on Windows with backslashes as separators and a drive letter
- All overwrite files are now created with explicit read-write 0o666, as the default 0o444 creates read-only files, leading to PHP Warnings `Failed to open`

---------

Co-authored-by: Allon Moritz <allon.moritz@digital-peak.com>
Co-authored-by: Robert Deutz <rdeutz@googlemail.com>

chmst committed:
[5.3] Remove unsed variable messages from enqueue messages (#42948)

* Remove unsed variable

* fix as reuested

* Add a comment

---------

Co-authored-by: Allon Moritz <allon.moritz@digital-peak.com>
Co-authored-by: Benjamin Trenkle <bembelimen@users.noreply.github.com>

richard67 committed:
Merge remote-tracking branch 'upstream/5.2-dev' into 5.3-dev-upmerge-2025-02-28

richard67 committed:
Merge pull request #45034 from richard67/5.3-dev-upmerge-2025-02-28

webgras committed:
Fix mail background colour and line for Outlook (#44985)

joomdonation committed:
[5.3] Remove useless sprintf function call (#45036)

alikon committed:
[5.3] fine grain task notification mail (#44604)

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Fedik committed:
[5.3] CMSPlugin: Deprecate use of DispatcherAware and LanguageAware (#43430)

joomdonation committed:
Fix code logic in admin controllers (#45037)

voronkovich committed:
Merge branch '5.3-dev' into config-envs

joomdonation committed:
Remove un-used import (#45042)

SniperSister committed:
[5.3] Fix core update information retrieval after changing the update channel or stability options (#44954)

voronkovich committed:
Merge branch '5.3-dev' into config-envs

voronkovich committed:
Fix setting JDEBUG for installation

voronkovich committed:
Tweak env.dist

brianteeman commented on env.dist:
```diff
@@ -0,0 +1,29 @@
+# JOOMLA_OFFLINE=1
+# JOOMLA_SITENAME='Joomla!'
+#
+# All available configuration values are listed in installation/configuration.php-dist.
```
is it ok to reference this file as it is removed once a site has been installed? Should it maybe refer to a documentation page that is always available?

brianteeman commented on tests/Unit/Libraries/Cms/Service/Provider/ConfigTest.php:
```diff
@@ -0,0 +1,93 @@
+ * @package     Joomla.UnitTest
+ * @subpackage  Service
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2025 Open Source Matters, Inc. <https://www.joomla.org>
```

voronkovich committed:
Update copyright year to 2025

Co-authored-by: Brian Teeman <brian@teeman.net>

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

voronkovich commented:
I accidentally removed the branch while rebasing and the PR closed automatically. I will try to figure out how to reopen it later.
